### DISCUSSION TITLE
Service with "<UNKNOWN_IDENTIFIER>" identifier was not found in the container.

### DISCUSSION THREAD

**Discussion by RyannGalea** [MichalLytek/type-graphql#978] [Q&A]
Previously when I was developing with **ts-node-dev**, I had little issues but now I've put together a build system to get my code to a production build I'm having some issues.

Consider that my code is written in Typescript and then uses Babel to transpile from Typescript to esnext, it then transpiles again down to es2015.

I never had an issue before but now I'm having this error for my HasRole middleware, and I have no direction on why this would now be an issue.

```
@UseMiddleware(HasRole([UserRolesEnum.ADMIN, UserRolesEnum.BOOKING_READ], RoleMessagesEnum.BOOKING_READ))
```
```
export const HasRole: (rolesArr: UserRolesEnum[], message?: string) => MiddlewareFn<Context> = (rolesArr: UserRolesEnum[], message?: string) => async ({ context: {req} }, next) => {

  // Get Roles
  const { roles } = req.session.user  
  if (!roles) throw new ApolloError(message || 'You do not have sufficient roles.') 

  // Check roles
  if (!(roles.some(str => rolesArr.includes(str)))) {
    throw new ApolloError(message || 'You do not have sufficient roles.')
  }

  return next()

}  
```
```
"Service with "<UNKNOWN_IDENTIFIER>" identifier was not found in the container. Register it before usage via explicitly calling the "Container.set" function or using the "@Service()" decorator.]"
```

Firstly is there a way to remedy this directly? This would be the easiest to keep the code as-is. Obviously, I can create class-based middleware but I'm not sure why I should only have to now, is this a TypeDI issue?

I already had to turn my global middleware into a class middleware so I could then decorate it with @Service but here it's a little bit harder because I'd like to just send the roles to the middleware directly, which is why I'm here asking for help on this.

The only way I can see doing it is creating class middleware which then has all the roles for each mutation in another service I can inject but I'd prefer to not go that direction.

Any guidance or help is appreciated.

Ryann,

**Comment by MichalLytek** [comment]
I bet this is an issue with your weird TS-babel-esnext-es2015 transpile chain, so it treat arrow function as normal function with prototype, thus detecting it as a class and tries to get the instance from container.

**Reply by RyannGalea** [reply]
ok, I'll look further in that direction.

I do believe there is an issue with the build steps, but struggling to find ways to resolve it

**Reply by RyannGalea** [reply]
Ok,

Seems that adding the targets option in Babel is helping for now.

```
    ["@babel/preset-env", { "targets": {"node": "14"}}],
```
